src/main/java/de/hpi/bpt/argos/persistence/model/parsing/BackboneDataParserImpl.java

```diff
@@ -0,0 +1,431 @@
+	protected static final String PRODUCT_ELEMENT = "product";
+	protected static final String PRODUCT_IDENTIFIER_ELEMENT = "productIdentifier";
+	protected static final String PRODUCT_DESCRIPTION_ELEMENT = "productDescription";
+	protected static final String PRODUCT_CONFIGURATION_ELEMENT = "productConfiguration";
+	protected static final String CODING_PLUG_ELEMENT = "codingPlug";
+	protected static final String CODING_PLUG_SOFTWARE_VERSION_ELEMENT = "codingPlugSoftwareVersion";
+	protected static final String DISPLAY_CODE_ELEMENT = "displayCode";
+	protected static final String CAUSE_CODE_ELEMENT = "causeCode";
+	protected static final String ERROR_DESCRIPTION_ELEMENT = "errorDescription";
+	protected static final String CAUSE_DESCRIPTION_ELEMENT = "causeDescription";
+	protected static final String CAUSE_PREDICTION_ELEMENT = "causePrediction";
+
+	protected static final String PRODUCT_FAMILY_SEPARATOR = "-";
+	protected static final int PRODUCT_FAMILY_SPLIT_LENGTH = 2;
+	protected static final double TO_PERCENT = 1.0 / 100.0;
+
+	protected List<Long> cachedExternalProductIdentifiers = new ArrayList<>();
+	protected String tempCurrentProductExternalId = "";
+	protected String tempCurrentProductFamilyId = "";
+	protected String tempCurrentCauseDescription = "";
+	protected String tempCurrentProductName = "";
+	protected String tempCurrentCauseCode = "";
+	protected String tempCurrentErrorDisplayCode = "";
+	protected String tempCurrentCodingPlug = "";
+	protected Product currentProduct = null;
+	protected ErrorType currentErrorType = null;
+	protected ProductConfiguration currentProductConfiguration = null;
+
+	// statistics
+	protected String fileName;
+	protected Date importStartTime;
+	protected long productsImported = 0;
+	protected long productConfigurationsImported = 0;
+	protected long errorTypesImported = 0;
+	protected long errorCausesImported = 0;
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void parse(File dataFile) {
+		fileName = dataFile.getName();
+		importStartTime = new Date();
+
+		List<ProductFamily> families = entityManager.getProductFamilies();
+		cachedExternalProductIdentifiers.clear();
+
+		for (ProductFamily family : families) {
+			for (Product product : family.getProducts()) {
+				cachedExternalProductIdentifiers.add(product.getOrderNumber());
+			}
+		}
+
+		super.parse(dataFile);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void endElement(String elementName) {
+		super.startElement(elementName);
+
+		switch (elementName) {
+			case PRODUCT_ELEMENT:
+				saveCurrentProduct();
+				break;
+
+			case PRODUCT_CONFIGURATION_ELEMENT:
+				currentProductConfiguration = null;
+				break;
+
+			case PRODUCTS_ELEMENT:
+				logStatistics();
+				break;
+
+			default:
+				// empty, since nothing to do
+				break;
+		}
+
+		if (elementName.equals(PRODUCT_ELEMENT)) {
+			saveCurrentProduct();
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void element(String elementData) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "element" is 11 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)